src/test/java/com/flightstats/hub/webhook/WebhookValidatorTest.java

```diff
@@ -3,158 +3,151 @@
-        when(appProperties.getHubType()).thenReturn("aws");
-        when(webhookProperties.getCallbackTimeoutMinimum()).thenReturn(1);
-        when(webhookProperties.getCallbackTimeoutMaximum()).thenReturn(1800);
-        webhook = webhook.withCallbackTimeoutSeconds(1000).withBatch("SINGLE").withName("blah");
-        webhookValidator.validate(webhook);
+        WebhookValidator webhookValidator = new WebhookValidator(getDefaultAppProperties(), getDefaultWebhookProperties());
+        Webhook webhook = getWebhook(builder -> builder.batch("single"));
+        assertDoesNotThrow(() -> webhookValidator.validate(webhook));
     }
 
     @Test
     void testInvalidSingleHeartbeat() {
-        webhook = webhook.withBatch("SINGLE").withHeartbeat(true).withName("blah").withCallbackTimeoutSeconds(10);
+        WebhookValidator webhookValidator = new WebhookValidator(getDefaultAppProperties(), getDefaultWebhookProperties());
+        Webhook webhook = getWebhook(builder -> builder.batch("SINGLE").heartbeat(true));
         assertThrows(InvalidRequestException.class, () -> webhookValidator.validate(webhook));
     }
 
     @Test
     void testInvalidCallbackTimeout() {
-        when(webhookProperties.getCallbackTimeoutMinimum()).thenReturn(1);
-        when(webhookProperties.getCallbackTimeoutMaximum()).thenReturn(1800);
-        webhook = webhook.withCallbackTimeoutSeconds(10 * 1000).withBatch("SINGLE").withName("blah");
+        WebhookValidator webhookValidator = new WebhookValidator(getDefaultAppProperties(), getDefaultWebhookProperties());
+        Webhook webhook = getWebhook(builder -> builder.callbackTimeoutSeconds(MAXIMUM_CALLBACK_TIMEOUT_IN_SEC + 1));
         assertThrows(InvalidRequestException.class, () -> webhookValidator.validate(webhook));
     }
 
     @Test
     void testInvalidCallbackTimeoutZero() {
-        when(webhookProperties.getCallbackTimeoutMinimum()).thenReturn(1);
-        when(webhookProperties.getCallbackTimeoutMaximum()).thenReturn(1800);
-        webhook = webhook.withCallbackTimeoutSeconds(0).withBatch("SINGLE").withName("blah");
+        WebhookValidator webhookValidator = new WebhookValidator(getDefaultAppProperties(), getDefaultWebhookProperties());
+        Webhook webhook = getWebhook(builder -> builder.callbackTimeoutSeconds(0));
         assertThrows(InvalidRequestException.class, () -> webhookValidator.validate(webhook));
     }
 
     @Test
     void testInvalidLocalhost() {
-        webhook = Webhook.builder()
-                .callbackUrl("http:/localhost:8080/url")
+        WebhookValidator webhookValidator = new WebhookValidator(getDefaultAppProperties(), getDefaultWebhookProperties());
+        Webhook webhook = getWebhook(builder -> builder.callbackUrl("http://localhost:8080/url"));
+        assertThrows(InvalidRequestException.class, () -> webhookValidator.validate(webhook));
+    }
+
+    private Webhook getWebhook(Function<Webhook.WebhookBuilder, Webhook.WebhookBuilder> customizations) {
```

Comment:
This uses a `Function<WebhookBuilder, WebhookBuilder>` rather than using `with`ers (i.e. `getWebhook().withName("invalid N@ame")`) because there are a few fields that can't be overridden once the webhook is built. I wanted creation of the webhooks to be consistent even with those fields, and this solved that problem.

Reply:
How is function call different from building webhook directly? Like instead of  line 138   customizations.apply(webhookBuilder).build(); the call could end on line 137 .heartbeat(false).build()
How are they different?

Reply:
Hopefully I'm answering the right question here... There are a few fields that can't be changed after you call `build()`; I think those fields are `name`, `channelUrl`, and `callbackUrl`.
So if line 137 did `.heartbeat(false).build()`, I couldn't use `getWebhook().withName("invalid N@me")` in the test because there's no wither for name. The old implementation of tests for those fields used `Webhook.builder().name(...)...build` in the test to get around that.

Unfortunately, there was at least one test for channel url that was failing because one of the OTHER fields specified in the build statement wasn't valid. Using a function lets the test inject the new, invalid values for those fields while maintaining a single definition of what a valid template for a webhook looks like...and kinda makes it more obvious about which fields/values are important to the test.

Sorry for writing so much there. Does that answer your question?

Reply:
That explains it. Thanks for the write up :)